ql/src/java/org/apache/hadoop/hive/ql/parse/rewrite/CopyOnWriteUpdateRewriter.java

```diff
@@ -45,40 +45,51 @@ public class CopyOnWriteUpdateRewriter implements Rewriter<UpdateStatement> {
   private final SetClausePatcher setClausePatcher;
 
 
-  public CopyOnWriteUpdateRewriter(HiveConf conf, SqlGeneratorFactory sqlGeneratorFactory,
-                                   COWWithClauseBuilder cowWithClauseBuilder, SetClausePatcher setClausePatcher) {
+  public CopyOnWriteUpdateRewriter(HiveConf conf, SqlGeneratorFactory sqlGeneratorFactory) {
     this.conf = conf;
     this.sqlGeneratorFactory = sqlGeneratorFactory;
-    this.cowWithClauseBuilder = cowWithClauseBuilder;
-    this.setClausePatcher = setClausePatcher;
+    this.cowWithClauseBuilder = new COWWithClauseBuilder();
+    this.setClausePatcher = new SetClausePatcher();
   }
 
   @Override
   public ParseUtils.ReparseResult rewrite(Context context, UpdateStatement updateBlock)
       throws SemanticException {
 
-    Tree wherePredicateNode = updateBlock.getWhereTree().getChild(0);
-    String whereClause = context.getTokenRewriteStream().toString(
-        wherePredicateNode.getTokenStartIndex(), wherePredicateNode.getTokenStopIndex());
     String filePathCol = HiveUtils.unparseIdentifier(VirtualColumn.FILE_PATH.getName(), conf);
-
     MultiInsertSqlGenerator sqlGenerator = sqlGeneratorFactory.createSqlGenerator();
 
-    cowWithClauseBuilder.appendWith(sqlGenerator, filePathCol, whereClause);
-
-    sqlGenerator.append("insert into table ");
+    String whereClause = null;
+    int columnOffset = 0;
+    
+    boolean shouldOverwrite = updateBlock.getWhereTree() == null;
```

Comment:
1. I think it would be better to move this to another Jira and patch since this isn't really about Merge.
2. I would like to avoid these type of flags in the implementation. Why don't we give another implementation of `Rewriter<UpdateStatement>`? It can extend `CopyOnWriteUpdateRewriter` if makes sense or vice versa. Or the common parts could be extracted to an abstract base class or into a wrapped class similar to `COWWithClauseBuilder` but with a different name like `CopyOnWriteUpdaterBase`

Reply:
any suggestions for the class name: CopyOnWriteIOWRewriter?

Reply:
I would prefer to keep the IOW optimization in COWUpdateRewriter in that PR, however, in follow-up PR we could move it to UpdateSemanticAnalyzer and use the same optimization for ACID

Reply:
created ticket [HIVE-27902](https://issues.apache.org/jira/browse/HIVE-27902)